File: js/src/so.js
"use strict";

const fs = require("fs");
const path = require("path");
const { soObjectPaths } = require("./signing");
const { HomeCloudError } = require("./errors");

const MIME_BY_EXT = {
  ".mp4": "video/mp4",
  ".webm": "video/webm",
  ".mov": "video/quicktime",
  ".mp3": "audio/mpeg",
  ".wav": "audio/wav",
  ".ogg": "audio/ogg",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".webp": "image/webp",
  ".pdf": "application/pdf",
  ".json": "application/json",
  ".txt": "text/plain",
  ".html": "text/html",
  ".htm": "text/html",
  ".css": "text/css",
  ".js": "text/javascript",
  ".xml": "application/xml",
  ".csv": "text/csv",
  ".zip": "application/zip",
};

function _guessContentType(name) {
  const ext = path.extname(String(name || "")).toLowerCase();
  return MIME_BY_EXT[ext] || null;
}

class SoAPI {
  constructor(client) {
    this._c = client;
  }

  async listBuckets() {
    this._c.requireConsole();
    const data = await this._c.consoleRequest(
      "GET",
      `accounts/${this._c.accountId}/storage/buckets`
    );
    return data.items || [];
  }

  async createBucket(name) {
    this._c.requireConsole();
    return this._c.consoleRequest("POST", `accounts/${this._c.accountId}/storage/buckets`, {
      json: { name: String(name).trim().toLowerCase() },
    });
  }

  async deleteBucket(name) {
    this._c.requireConsole();
    await this._c.consoleRequest(
      "DELETE",
      `accounts/${this._c.accountId}/storage/buckets/${String(name).trim().toLowerCase()}`
    );
  }

  async listObjects(bucketName, { prefix = "", recursive = false, page = 1, pageSize = 100 } = {}) {
    this._c.requireAccessKey();
    const accountId = this._c.accountId;
    const reqPath = `/${accountId}/${bucketName}/objects`;
    return this._c.dataPlaneRequest("so", "GET", reqPath, {
      params: { prefix, recursive, page, page_size: pageSize },
    });
  }

  async listAllObjects(bucketName, { prefix = "", recursive = true } = {}) {
    const items = [];
    let page = 1;
    for (;;) {
      const data = await this.listObjects(bucketName, { prefix, recursive, page, pageSize: 100 });
      for (const item of data.items || []) {
        if (!item.is_dir) items.push(item);
      }
      if (page >= Number(data.pages || 1)) break;
      page += 1;
    }
    return items;
  }

  async upload(bucketName, filePath, { key, body, contentType } = {}) {
    this._c.requireAccessKey();
    const hasBody = body !== undefined && body !== null;
    if (hasBody && filePath != null && filePath !== "") {
      throw new HomeCloudError("Pass either filePath or body, not both");
    }
    if (!hasBody && (filePath == null || filePath === "")) {
      throw new HomeCloudError("filePath or body is required");
    }
    if (hasBody && (!key || !String(key).trim())) {
      throw new HomeCloudError("key is required when uploading body");
    }

    let objectKey;
    let filename;
    let payload;

    if (hasBody) {
      objectKey = String(key).trim().replace(/^\/+/, "");
      filename = path.basename(objectKey) || "object";
      payload = body;
    } else {
      if (!fs.existsSync(filePath) || !fs.statSync(filePath).isFile()) {
        throw new HomeCloudError(`File not found: ${filePath}`);
      }
      objectKey = key || path.basename(filePath);
      filename = path.basename(filePath);
      payload = fs.readFileSync(filePath);
    }

    const mime =
      contentType ||
      _guessContentType(objectKey) ||
      _guessContentType(filename) ||
      "application/octet-stream";
    const accountId = this._c.accountId;
    const uploadPath = `/${accountId}/${bucketName}/objects`;
    const form = new FormData();
    form.append("key", objectKey);
    form.append("file", new Blob([payload], { type: mime }), filename);
    return this._c.dataPlaneRequest("so", "POST", uploadPath, { formData: form });
  }

  async putJson(bucketName, objectKey, value) {
    const data = Buffer.from(JSON.stringify(value, null, 2), "utf8");
    return this.upload(bucketName, null, {
      key: objectKey,
      body: data,
      contentType: "application/json",
    });
  }

  async delete(bucketName, objectKey) {
    this._c.requireAccessKey();
    const { signPath, urlPath } = soObjectPaths(this._c.accountId, bucketName, objectKey);
    await this._c.dataPlaneRequest("so", "DELETE", signPath, { urlPath, signPath });
  }

  async download(bucketName, objectKey, { destPath } = {}) {
    this._c.requireAccessKey();
    if (!destPath) throw new HomeCloudError("destPath is required");
    const key = String(objectKey).replace(/^\/+/, "");
    const { signPath, urlPath } = soObjectPaths(this._c.accountId, bucketName, key);
    const buf = await this._c.dataPlaneRequestBytes("so", "GET", signPath, { urlPath, signPath });
    fs.mkdirSync(path.dirname(destPath), { recursive: true });
    fs.writeFileSync(destPath, buf);
    return { key, size: buf.length, path: destPath };
  }

  async headObject(bucketName, objectKey) {
    this._c.requireAccessKey();
    const key = String(objectKey).replace(/^\/+/, "");
    const { signPath, urlPath } = soObjectPaths(this._c.accountId, bucketName, key);
    const raw = await this._c.dataPlaneRequest("so", "GET", `${signPath}/metadata`, {
      urlPath: `${urlPath}/metadata`,
      signPath: `${signPath}/metadata`,
    });
    if (!raw || typeof raw !== "object") throw new HomeCloudError("Invalid metadata response");
    const userMeta = raw.metadata && typeof raw.metadata === "object" ? raw.metadata : {};
    const tags = raw.tags && typeof raw.tags === "object" ? raw.tags : {};
    return {
      key: String(raw.key || key),
      size: Number(raw.size || 0),
      etag: raw.etag,
      content_type: raw.content_type,
      last_modified: raw.last_modified,
      metadata: Object.fromEntries(Object.entries(userMeta).map(([k, v]) => [String(k), String(v)])),
      tags: Object.fromEntries(Object.entries(tags).map(([k, v]) => [String(k), String(v)])),
    };
  }

  async objectMetadata(bucketName, objectKey) {
    return this.headObject(bucketName, objectKey);
  }

  async getObjectUri(bucketName, objectKey) {
    this._c.requireAccessKey();
    const key = String(objectKey).replace(/^\/+/, "");
    const { signPath, urlPath } = soObjectPaths(this._c.accountId, bucketName, key);
    const raw = await this._c.dataPlaneRequest("so", "GET", `${signPath}/uri`, {
      urlPath: `${urlPath}/uri`,
      signPath: `${signPath}/uri`,
    });
    if (!raw || typeof raw !== "object") throw new HomeCloudError("Invalid URI response");
    return {
      so_uri: String(raw.so_uri || `so://${bucketName}/${key}`),
      https_url: String(raw.https_url || ""),
      https_requires_public: Boolean(raw.https_requires_public ?? true),
    };
  }

  async generatePresignedUrl(bucketName, objectKey, { expires = 3600 } = {}) {
    this._c.requireAccessKey();
    const key = String(objectKey).replace(/^\/+/, "");
    const { signPath, urlPath } = soObjectPaths(this._c.accountId, bucketName, key);
    const raw = await this._c.dataPlaneRequest("so", "GET", `${signPath}/presigned`, {
      urlPath: `${urlPath}/presigned`,
      signPath: `${signPath}/presigned`,
      params: { expires },
    });
    if (!raw || !raw.url) throw new HomeCloudError("Invalid presigned URL response");
    return {
      url: String(raw.url),
      expires_in_seconds: Number(raw.expires_in_seconds || expires),
    };
  }

  async copy(bucketName, sourceKey, destinationKey, { sourceBucket = null } = {}) {
    this._c.requireAccessKey();
    const { signPath, urlPath } = soObjectPaths(this._c.accountId, bucketName, sourceKey);
    return this._c.dataPlaneRequest("so", "POST", `${signPath}/copy`, {
      urlPath: `${urlPath}/copy`,
      signPath: `${signPath}/copy`,
      json: {
        destination_key: destinationKey,
        source_bucket: sourceBucket || null,
      },
    });
  }

  async deleteRecursive(bucketName, prefix = "") {
    const items = await this.listAllObjects(bucketName, { prefix, recursive: true });
    for (const item of items) {
      await this.delete(bucketName, item.key);
    }
    return items.length;
  }

  _isSoUri(target) {
    const lowered = String(target || "").toLowerCase();
    return lowered.startsWith("so://") || lowered.startsWith("s3://");
  }

  _parseSoUri(target) {
    let text = String(target || "").trim();
    const lowered = text.toLowerCase();
    if (lowered.startsWith("so://")) text = text.slice(5);
    else if (lowered.startsWith("s3://")) text = text.slice(5);
    text = text.replace(/^\/+|\/+$/g, "");
    if (!text) throw new HomeCloudError("URI must include a bucket name");
    const slash = text.indexOf("/");
    if (slash < 0) return { bucket: text, prefix: "" };
    return { bucket: text.slice(0, slash), prefix: text.slice(slash + 1) };
  }

  _syncJoinPrefix(prefixClean, relative) {
    const rel = String(relative || "").replace(/^\/+/, "");
    if (!prefixClean) return rel;
    if (!rel) return prefixClean;
    return `${prefixClean}/${rel}`;
  }

  _syncRelativePath(key, prefixClean) {
    if (!prefixClean) return key;
    if (key === prefixClean) return key.split("/").pop();
    if (key.startsWith(`${prefixClean}/`)) return key.slice(prefixClean.length + 1);
    return key;
  }

  /**
   * Unified sync: local↔bucket or bucket↔bucket.
   * Prefer this over syncLocalToBucket / syncBucketToLocal.
   */
  async sync(source, destination, { deleteExtra = false, skip = false } = {}) {
    const srcRemote = this._isSoUri(source);
    const dstRemote = this._isSoUri(destination);
    if (srcRemote && dstRemote) {
      const src = this._parseSoUri(source);
      const dst = this._parseSoUri(destination);
      return this._syncBucketToBucket(src.bucket, dst.bucket, {
        sourcePrefix: src.prefix,
        destinationPrefix: dst.prefix,
        deleteExtra,
        skip,
      });
    }
    if (srcRemote && !dstRemote) {
      const src = this._parseSoUri(source);
      return this.syncBucketToLocal(src.bucket, destination, {
        prefix: src.prefix,
        deleteExtra,
        skip,
      });
    }
    if (!srcRemote && dstRemote) {
      const dst = this._parseSoUri(destination);
      return this.syncLocalToBucket(source, dst.bucket, {
        prefix: dst.prefix,
        deleteExtra,
        skip,
      });
    }
    throw new HomeCloudError(
      "One or both sides must be an so:// URI (local↔bucket or bucket↔bucket)"
    );
  }

  async _syncBucketToBucket(
    sourceBucket,
    destinationBucket,
    { sourcePrefix = "", destinationPrefix = "", deleteExtra = false, skip = false } = {}
  ) {
    this._c.requireAccessKey();
    const srcPrefix = String(sourcePrefix || "").replace(/^\/+|\/+$/g, "");
    const dstPrefix = String(destinationPrefix || "").replace(/^\/+|\/+$/g, "");
    if (sourceBucket === destinationBucket && srcPrefix === dstPrefix) {
      throw new HomeCloudError(`Source and destination are the same: so://${sourceBucket}/${srcPrefix}`);
    }

    const sourceItems = await this.listAllObjects(sourceBucket, {
      prefix: srcPrefix,
      recursive: true,
    });
    const destItems = await this.listAllObjects(destinationBucket, {
      prefix: dstPrefix,
      recursive: true,
    });

    const sourceRels = new Map();
    for (const item of sourceItems) {
      const rel = this._syncRelativePath(item.key, srcPrefix);
      sourceRels.set(rel, { key: item.key, size: Number(item.size || 0) });
    }
    const destRels = new Map();
    for (const item of destItems) {
      const rel = this._syncRelativePath(item.key, dstPrefix);
      destRels.set(rel, { key: item.key, size: Number(item.size || 0) });
    }

    let copied = 0;
    let skipped = 0;
    for (const [rel, src] of sourceRels) {
      const dest = destRels.get(rel);
      if (skip && dest && dest.size === src.size) {
        skipped += 1;
        continue;
      }
      const dstKey = this._syncJoinPrefix(dstPrefix, rel);
      await this.copy(destinationBucket, src.key, dstKey, {
        sourceBucket: sourceBucket !== destinationBucket ? sourceBucket : null,
      });
      copied += 1;
    }

    let deleted = 0;
    if (deleteExtra) {
      for (const [rel, dest] of destRels) {
        if (sourceRels.has(rel)) continue;
        await this.delete(destinationBucket, dest.key);
        deleted += 1;
      }
    }
    return { copied, skipped, deleted };
  }

  /** Prefer sync("./dir", "so://bucket/prefix"). */
  async syncLocalToBucket(localDir, bucketName, { prefix = "", deleteExtra = false, skip = false } = {}) {
    this._c.requireAccessKey();
    const root = path.resolve(localDir);
    const prefixClean = String(prefix || "").replace(/^\/+|\/+$/g, "");
    const walk = (dir, base) => {
      const out = [];
      for (const name of fs.readdirSync(dir)) {
        const full = path.join(dir, name);
        const st = fs.statSync(full);
        const rel = path.relative(base, full).split(path.sep).join("/");
        if (st.isDirectory()) out.push(...walk(full, base));
        else out.push({ rel, size: st.size });
      }
      return out;
    };
    const locals = walk(root, root);
    const remote = await this.listAllObjects(bucketName, { prefix: prefixClean, recursive: true });
    const remoteByKey = new Map(remote.map((item) => [item.key, item]));

    let uploaded = 0;
    let skipped = 0;
    for (const { rel, size } of locals) {
      const key = this._syncJoinPrefix(prefixClean, rel);
      const existing = remoteByKey.get(key);
      if (skip && existing && Number(existing.size || 0) === size) {
        skipped += 1;
        continue;
      }
      await this.upload(bucketName, path.join(root, rel), { key });
      uploaded += 1;
    }
    let deleted = 0;
    if (deleteExtra) {
      const localKeys = new Set(locals.map(({ rel }) => this._syncJoinPrefix(prefixClean, rel)));
      for (const item of remote) {
        if (!localKeys.has(item.key)) {
          await this.delete(bucketName, item.key);
          deleted += 1;
        }
      }
    }
    return { uploaded, skipped, deleted };
  }

  /** Prefer sync("so://bucket/prefix", "./dir"). */
  async syncBucketToLocal(bucketName, localDir, { prefix = "", deleteExtra = false, skip = false } = {}) {
    this._c.requireAccessKey();
    const root = path.resolve(localDir);
    const prefixClean = String(prefix || "").replace(/^\/+|\/+$/g, "");
    const items = await this.listAllObjects(bucketName, { prefix: prefixClean, recursive: true });
    let downloaded = 0;
    let skippedCount = 0;
    const remoteRels = new Set();
    for (const item of items) {
      const rel = this._syncRelativePath(item.key, prefixClean);
      remoteRels.add(rel);
      const dest = path.join(root, rel);
      if (
        skip &&
        fs.existsSync(dest) &&
        fs.statSync(dest).isFile() &&
        fs.statSync(dest).size === Number(item.size || 0)
      ) {
        skippedCount += 1;
        continue;
      }
      await this.download(bucketName, item.key, { destPath: dest });
      downloaded += 1;
    }
    let deleted = 0;
    if (deleteExtra && fs.existsSync(root)) {
      const walk = (dir, base) => {
        const out = [];
        for (const name of fs.readdirSync(dir)) {
          const full = path.join(dir, name);
          const st = fs.statSync(full);
          if (st.isDirectory()) out.push(...walk(full, base));
          else out.push(path.relative(base, full).split(path.sep).join("/"));
        }
        return out;
      };
      for (const rel of walk(root, root)) {
        if (remoteRels.has(rel)) continue;
        fs.unlinkSync(path.join(root, rel));
        deleted += 1;
      }
    }
    return { downloaded, skipped: skippedCount, deleted };
  }
}

module.exports = { SoAPI };

File: java/src/main/java/com/homecloudlab/sdk/So.java
package com.homecloudlab.sdk;

import com.fasterxml.jackson.databind.JsonNode;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/** Object Storage — data plane plus console bucket helpers. */
public final class So {
    private final HomeCloud c;

    So(HomeCloud c) {
        this.c = c;
    }

    public List<Bucket> listBuckets() {
        if (c.hasAccessKey()) {
            c.ensureAccountId();
            byte[] raw = c.dataPlaneJson("so", "GET", "/" + c.accountIdOrEmpty() + "/buckets", "", null, null, null);
            return Json.itemsOf(raw, Bucket.class);
        }
        c.requireConsole();
        c.ensureAccountId();
        byte[] raw = c.consoleJson("GET", "accounts/" + c.accountIdOrEmpty() + "/storage/buckets", true, null, null, null, null);
        return Json.itemsOf(raw, Bucket.class);
    }

    public Bucket createBucket(String name) {
        c.ensureAccountId();
        byte[] raw = c.consoleJson(
                "POST",
                "accounts/" + c.accountIdOrEmpty() + "/storage/buckets",
                true,
                Map.of("name", name.trim().toLowerCase(Locale.ROOT)),
                null,
                Transport.newIdempotencyKey(),
                Transport.RetryMode.IF_IDEMPOTENCY);
        return Json.decode(raw, Bucket.class);
    }

    public void deleteBucket(String name) {
        c.ensureAccountId();
        c.consoleJson(
                "DELETE",
                "accounts/" + c.accountIdOrEmpty() + "/storage/buckets/" + name.trim().toLowerCase(Locale.ROOT),
                true, null, null, null, null);
    }

    public ListObjectsResult listObjects(String bucket, ListObjectsOptions opts) {
        c.requireAccessKey();
        c.ensureAccountId();
        if (opts == null) {
            opts = ListObjectsOptions.builder().build();
        }
        int page = opts.page() <= 0 ? 1 : opts.page();
        int pageSize = opts.pageSize() <= 0 ? 100 : opts.pageSize();
        Map<String, String> q = new LinkedHashMap<>();
        q.put("prefix", opts.prefix());
        q.put("recursive", opts.recursive() ? "true" : "false");
        q.put("page", Integer.toString(page));
        q.put("page_size", Integer.toString(pageSize));
        if (!opts.continuationToken().isEmpty()) {
            q.put("continuation_token", opts.continuationToken());
        }
        String path = "/" + c.accountIdOrEmpty() + "/" + bucket + "/objects";
        byte[] raw = c.dataPlaneJson("so", "GET", path, "", q, null, null);
        ListObjectsResult res = Json.decode(raw, ListObjectsResult.class);
        return res == null ? new ListObjectsResult(List.of(), false, "", null) : res;
    }

    public List<ObjectListItem> listAllObjects(String bucket, ListObjectsOptions opts) {
        if (opts == null) {
            opts = ListObjectsOptions.builder().build();
        }
        List<ObjectListItem> items = new ArrayList<>();
        int page = 1;
        String token = "";
        boolean recursive = true;
        if (opts.page() > 0 || opts.pageSize() > 0) {
            recursive = opts.recursive();
        }
        while (true) {
            ListObjectsResult data = listObjects(bucket, ListObjectsOptions.builder()
                    .prefix(opts.prefix())
                    .recursive(recursive)
                    .page(page)
                    .pageSize(100)
                    .continuationToken(token)
                    .build());
            for (ObjectListItem item : data.items()) {
                if (!item.isDir()) {
                    items.add(item);
                }
            }
            if (data.hasMore() && data.nextContinuationToken() != null && !data.nextContinuationToken().isEmpty()) {
                token = data.nextContinuationToken();
                page = 1;
                continue;
            }
            if (data.pages() != null && page < data.pages()) {
                page++;
                token = "";
                continue;
            }
            break;
        }
        return items;
    }

    public ObjectRef upload(String bucket, UploadOptions opts) {
        c.requireAccessKey();
        c.ensureAccountId();
        if (opts.body() != null && opts.filePath() != null && !opts.filePath().isEmpty()) {
            throw new HomeCloudException("Pass either filePath or body, not both");
        }
        if (opts.body() == null && (opts.filePath() == null || opts.filePath().isEmpty())) {
            throw new HomeCloudException("filePath or body is required");
        }
        if (opts.body() != null && (opts.key() == null || opts.key().isBlank())) {
            throw new HomeCloudException("key is required when uploading body");
        }
        Transport.Spec spec = new Transport.Spec();
        spec.method = "POST";
        spec.accountId = c.accountIdOrEmpty();
        spec.signed = true;
        spec.retry = Transport.RetryMode.UPLOAD;
        String objectKey;
        String filename;
        if (opts.body() != null) {
            objectKey = SoPaths.trimLeftSlash(opts.key().trim());
            filename = Path.of(objectKey).getFileName() == null ? "object" : Path.of(objectKey).getFileName().toString();
            if (filename.isEmpty() || ".".equals(filename)) {
                filename = "object";
            }
            spec.multipartBytes = opts.body();
        } else {
            Path p = Path.of(opts.filePath());
            if (!Files.isRegularFile(p)) {
                throw new HomeCloudException("File not found: " + opts.filePath());
            }
            objectKey = opts.key() == null || opts.key().isBlank() ? p.getFileName().toString() : opts.key();
            filename = p.getFileName().toString();
            spec.multipartFile = opts.filePath();
        }
        spec.multipartKey = objectKey;
        spec.multipartName = filename;
        spec.signPath = "/" + c.accountIdOrEmpty() + "/" + bucket + "/objects";
        spec.url = c.dataPlaneBase("so") + spec.signPath;
        byte[] raw = Transport.doRequest(c, spec);
        ObjectRef ref = Json.decode(raw, ObjectRef.class);
        if (ref == null || ref.key() == null || ref.key().isEmpty()) {
            return new ObjectRef(objectKey, ref == null ? "" : ref.etag(), ref == null ? 0 : ref.size());
        }
        return ref;
    }

    public ObjectRef putJson(String bucket, String key, Object value) {
        try {
            byte[] raw = Json.MAPPER.writerWithDefaultPrettyPrinter().writeValueAsBytes(value);
            return upload(bucket, UploadOptions.builder().key(key).body(raw).contentType("application/json").build());
        } catch (HomeCloudException e) {
            throw e;
        } catch (Exception e) {
            throw new HomeCloudException("Invalid JSON", e);
        }
    }

    public void delete(String bucket, String objectKey) {
        c.ensureAccountId();
        String[] paths = SoPaths.soObjectPaths(c.accountIdOrEmpty(), bucket, objectKey);
        c.dataPlaneJson("so", "DELETE", paths[0], paths[1], null, null, null);
    }

    public JsonNode copy(String bucket, String sourceKey, String destinationKey, CopyOptions opts) {
        c.ensureAccountId();
        String[] paths = SoPaths.soObjectPaths(c.accountIdOrEmpty(), bucket, sourceKey);
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("destination_key", destinationKey);
        body.put("source_bucket", opts != null && opts.sourceBucket() != null && !opts.sourceBucket().isEmpty()
                ? opts.sourceBucket() : null);
        return Json.parse(c.dataPlaneJson("so", "POST", paths[0] + "/copy", paths[1] + "/copy", null, body, Transport.RetryMode.UPLOAD));
    }

    public JsonNode move(String bucket, String sourceKey, String destinationKey, CopyOptions opts) {
        JsonNode copied = copy(bucket, sourceKey, destinationKey, opts);
        headObject(bucket, destinationKey);
        String srcBucket = opts != null && opts.sourceBucket() != null && !opts.sourceBucket().isEmpty()
                ? opts.sourceBucket() : bucket;
        delete(srcBucket, sourceKey);
        return copied;
    }

    public DownloadResult download(String bucket, String objectKey, String destPath) {
        c.requireAccessKey();
        c.ensureAccountId();
        String key = SoPaths.trimLeftSlash(objectKey);
        String[] paths = SoPaths.soObjectPaths(c.accountIdOrEmpty(), bucket, key);
        try {
            Path dest = Path.of(destPath);
            if (dest.getParent() != null) {
                Files.createDirectories(dest.getParent());
            }
            Transport.Spec spec = new Transport.Spec();
            spec.method = "GET";
            spec.url = c.dataPlaneBase("so") + paths[1];
            spec.signPath = paths[0];
            spec.accountId = c.accountIdOrEmpty();
            spec.signed = true;
            spec.retry = Transport.RetryMode.IDEMPOTENT;
            spec.stream = true;
            byte[] raw = Transport.doRequest(c, spec);
            Files.write(dest, raw);
            return new DownloadResult(key, raw.length, destPath);
        } catch (HomeCloudException e) {
            throw e;
        } catch (Exception e) {
            throw new HomeCloudException("Download failed", e);
        }
    }

    public ObjectHead headObject(String bucket, String objectKey) {
        c.ensureAccountId();
        String key = SoPaths.trimLeftSlash(objectKey);
        String[] paths = SoPaths.soObjectPaths(c.accountIdOrEmpty(), bucket, key);
        byte[] raw = c.dataPlaneJson("so", "GET", paths[0] + "/metadata", paths[1] + "/metadata", null, null, null);
        JsonNode parsed = Json.parse(raw);
        Map<String, Object> map = Json.asMap(parsed);
        String k = Errors.stringify(map.get("key"));
        ObjectHead head = new ObjectHead(
                k.isEmpty() ? key : k,
                jsonLong(map.get("size")),
                Errors.stringify(map.get("etag")),
                Errors.stringify(map.get("content_type")),
                Errors.stringify(map.get("last_modified")),
                Json.stringMap(map.get("metadata")),
                Json.stringMap(map.get("tags")));
        return head;
    }

    public ObjectUri getObjectUri(String bucket, String objectKey) {
        c.ensureAccountId();
        String key = SoPaths.trimLeftSlash(objectKey);
        String[] paths = SoPaths.soObjectPaths(c.accountIdOrEmpty(), bucket, key);
        byte[] raw = c.dataPlaneJson("so", "GET", paths[0] + "/uri", paths[1] + "/uri", null, null, null);
        ObjectUri u = Json.decode(raw, ObjectUri.class);
        if (u == null || u.soUri() == null || u.soUri().isEmpty()) {
            return new ObjectUri("so://" + bucket + "/" + key, u == null ? "" : u.httpsUrl(), u != null && u.httpsRequiresPublic());
        }
        return u;
    }

    public PresignedUrl generatePresignedUrl(String bucket, String objectKey, int expires) {
        if (expires <= 0) {
            expires = 3600;
        }
        c.ensureAccountId();
        String key = SoPaths.trimLeftSlash(objectKey);
        String[] paths = SoPaths.soObjectPaths(c.accountIdOrEmpty(), bucket, key);
        Map<String, String> q = Map.of("expires", Integer.toString(expires));
        byte[] raw = c.dataPlaneJson("so", "GET", paths[0] + "/presigned", paths[1] + "/presigned", q, null, null);
        PresignedUrl u = Json.decode(raw, PresignedUrl.class);
        if (u == null || u.url() == null || u.url().isEmpty()) {
            throw new HomeCloudException("Invalid presigned URL response");
        }
        if (u.expiresInSeconds() == 0) {
            return new PresignedUrl(u.url(), expires);
        }
        return u;
    }

    public int deleteRecursive(String bucket, String prefix) {
        List<ObjectListItem> items = listAllObjects(bucket, ListObjectsOptions.builder().prefix(prefix).recursive(true).build());
        for (ObjectListItem item : items) {
            delete(bucket, item.key());
        }
        return items.size();
    }

    public SyncResult sync(String source, String destination, SyncOptions opts) {
        if (opts == null) {
            opts = SyncOptions.none();
        }
        boolean srcRemote = SoPaths.isSoUri(source);
        boolean dstRemote = SoPaths.isSoUri(destination);
        if (srcRemote && dstRemote) {
            String[] s = SoPaths.parseSoUri(source);
            String[] d = SoPaths.parseSoUri(destination);
            return syncBucketToBucket(s[0], d[0], s[1], d[1], opts);
        }
        if (srcRemote) {
            String[] s = SoPaths.parseSoUri(source);
            return syncBucketToLocal(s[0], destination, s[1], opts);
        }
        if (dstRemote) {
            String[] d = SoPaths.parseSoUri(destination);
            return syncLocalToBucket(source, d[0], d[1], opts);
        }
        throw new HomeCloudException("One or both sides must be an so:// URI. Examples: ./dir so://bucket/  |  so://bucket/ ./dir  |  so://a/ so://b/");
    }

    private SyncResult syncLocalToBucket(String localDir, String bucket, String prefix, SyncOptions opts) {
        Path dir = Path.of(localDir);
        if (!Files.isDirectory(dir)) {
            throw new HomeCloudException("Not a directory: " + localDir);
        }
        String prefixClean = trimSlash(prefix);
        Map<String, Path> localFiles = new LinkedHashMap<>();
        try (var walk = Files.walk(dir)) {
            walk.filter(Files::isRegularFile).forEach(path -> {
                String rel = dir.relativize(path).toString().replace('\\', '/');
                localFiles.put(SoPaths.syncJoinPrefix(prefixClean, rel), path);
            });
        } catch (Exception e) {
            throw new HomeCloudException("Failed to walk directory", e);
        }
        Map<String, ObjectListItem> remoteByKey = new LinkedHashMap<>();
        for (ObjectListItem it : listAllObjects(bucket, ListObjectsOptions.builder().prefix(prefixClean).recursive(true).build())) {
            remoteByKey.put(it.key(), it);
        }
        int uploaded = 0, skipped = 0, deleted = 0;
        for (var e : localFiles.entrySet()) {
            try {
                long size = Files.size(e.getValue());
                if (opts.skip()) {
                    ObjectListItem r = remoteByKey.get(e.getKey());
                    if (r != null && r.size() == size) {
                        skipped++;
                        continue;
                    }
                }
                upload(bucket, UploadOptions.builder().filePath(e.getValue().toString()).key(e.getKey()).build());
                uploaded++;
            } catch (HomeCloudException ex) {
                throw ex;
            } catch (Exception ex) {
                throw new HomeCloudException("Sync upload failed", ex);
            }
        }
        if (opts.delete()) {
            for (String key : remoteByKey.keySet()) {
                if (!localFiles.containsKey(key)) {
                    delete(bucket, key);
                    deleted++;
                }
            }
        }
        return new SyncResult(uploaded, 0, 0, skipped, deleted);
    }

    private SyncResult syncBucketToLocal(String bucket, String localDir, String prefix, SyncOptions opts) {
        try {
            Files.createDirectories(Path.of(localDir));
        } catch (Exception e) {
            throw new HomeCloudException("Failed to create directory", e);
        }
        String prefixClean = trimSlash(prefix);
        int downloaded = 0, skipped = 0, deleted = 0;
        Map<String, ObjectListItem> remoteKeys = new LinkedHashMap<>();
        for (ObjectListItem it : listAllObjects(bucket, ListObjectsOptions.builder().prefix(prefixClean).recursive(true).build())) {
            remoteKeys.put(it.key(), it);
            String rel = SoPaths.syncRelativeLocalPath(it.key(), prefixClean);
            Path dest = Path.of(localDir, rel.replace('/', java.io.File.separatorChar));
            if (opts.skip() && Files.isRegularFile(dest)) {
                try {
                    if (Files.size(dest) == it.size()) {
                        skipped++;
                        continue;
                    }
                } catch (Exception ignored) {
                    // download
                }
            }
            download(bucket, it.key(), dest.toString());
            downloaded++;
        }
        if (opts.delete()) {
            Path root = Path.of(localDir);
            try (var walk = Files.walk(root)) {
                for (Path path : walk.filter(Files::isRegularFile).toList()) {
                    String rel = root.relativize(path).toString().replace('\\', '/');
                    String key = SoPaths.syncJoinPrefix(prefixClean, rel);
                    if (!remoteKeys.containsKey(key)) {
                        Files.deleteIfExists(path);
                        deleted++;
                    }
                }
            } catch (Exception e) {
                throw new HomeCloudException("Sync delete failed", e);
            }
        }
        return new SyncResult(0, downloaded, 0, skipped, deleted);
    }

    private SyncResult syncBucketToBucket(String srcBucket, String dstBucket, String srcPrefix, String dstPrefix, SyncOptions opts) {
        srcPrefix = trimSlash(srcPrefix);
        dstPrefix = trimSlash(dstPrefix);
        Map<String, ObjectListItem> destByRel = new LinkedHashMap<>();
        for (ObjectListItem it : listAllObjects(dstBucket, ListObjectsOptions.builder().prefix(dstPrefix).recursive(true).build())) {
            destByRel.put(SoPaths.syncRelativeLocalPath(it.key(), dstPrefix), it);
        }
        int copied = 0, skipped = 0, deleted = 0;
        Map<String, ObjectListItem> srcRels = new LinkedHashMap<>();
        for (ObjectListItem it : listAllObjects(srcBucket, ListObjectsOptions.builder().prefix(srcPrefix).recursive(true).build())) {
            String rel = SoPaths.syncRelativeLocalPath(it.key(), srcPrefix);
            srcRels.put(rel, it);
            if (opts.skip()) {
                ObjectListItem d = destByRel.get(rel);
                if (d != null && d.size() == it.size()) {
                    skipped++;
                    continue;
                }
            }
            String dstKey = SoPaths.syncJoinPrefix(dstPrefix, rel);
            CopyOptions copyOpts = srcBucket.equals(dstBucket) ? CopyOptions.none() : new CopyOptions(srcBucket);
            copy(dstBucket, it.key(), dstKey, copyOpts);
            copied++;
        }
        if (opts.delete()) {
            for (var e : destByRel.entrySet()) {
                if (!srcRels.containsKey(e.getKey())) {
                    delete(dstBucket, e.getValue().key());
                    deleted++;
                }
            }
        }
        return new SyncResult(0, 0, copied, skipped, deleted);
    }

    private static String trimSlash(String s) {
        if (s == null) {
            return "";
        }
        int a = 0, b = s.length();
        while (a < b && s.charAt(a) == '/') a++;
        while (b > a && s.charAt(b - 1) == '/') b--;
        return s.substring(a, b);
    }

    private static long jsonLong(Object v) {
        if (v instanceof Number n) {
            return n.longValue();
        }
        if (v == null) {
            return 0;
        }
        try {
            return Long.parseLong(String.valueOf(v));
        } catch (NumberFormatException e) {
            return 0;
        }
    }
}
